docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+
+ But we can further push this filter down to the inner table, since only the the records satisfy `t2.a in (12, 13)` could make join predicate `t1.a = t2.a` be positive in the join operator. So we can optimize this query to:
+
+ `select * from t1 left join t2 on t1.a=t2.a and t2.a in (12, 13) where t1.a in (12, 13);`
+
+And the join predicate `t2.a in (12, 13)` can be pushed down:
+
+```
+TiDB(localhost:4000) > desc select * from t1 left join t2 on t1.a=t2.a and t2.a in (12, 13) where t1.a in (12, 13);
++-------------------------+-------+------+-------------------------------------------------------------------------+
+| id                      | count | task | operator info                                                           |
++-------------------------+-------+------+-------------------------------------------------------------------------+
+| HashLeftJoin_7          | 0.00  | root | left outer join, inner:TableReader_13, equal:[eq(test.t1.a, test.t2.a)] |
+| ├─TableReader_10        | 0.00  | root | data:Selection_9                                                        |
+| │ └─Selection_9         | 0.00  | cop  | in(test.t1.a, 12, 13)                                                   |
+| │   └─TableScan_8       | 2.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo             |
+| └─TableReader_13        | 0.00  | root | data:Selection_12                                                       |
+|   └─Selection_12        | 0.00  | cop  | in(test.t2.a, 12, 13)                                                   |
+|     └─TableScan_11      | 2.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo             |
++-------------------------+-------+------+-------------------------------------------------------------------------+
+7 rows in set (0.00 sec)
+```
+
+## Rationale
+
+Constraint propagation is commonly used as logcial plan optimization in traditional databases, for example, [this doc](https://dev.mysql.com/doc/internals/en/optimizer-constant-propagation.html) explained some details of constant propagtions in MySQL. It is is also widely adopted in distributed analytical engines, like Apache Hive, Apache SparkSQL, Apache Impala, et al, those engines usually query on huge amount of data.
+
+### Advantages:
+
+ Constraint propagation brings more detailed, explicit constraints to each data source involved in a query. With those constraints we can filter data as early as possible, and thus reduce disc/network IO and computational overheads during the execution of a query. In TiDB, most propagated filters can be pushed down to the storage level(TiKV) as a coprocessor task, and lead to the following benefits:
+
+ * Apply the filters at each TiKV instance, which make the calculation distributed.
+
+ * When loading data, skip some table partitions if its data range doesn't pass the filter
```

Comment:
some table or partitions? Those are two different things.

Reply:
This was ambiguous, I was trying to say 'skip some partitions of table if the partitioning expression doesn't pass the filter', it is for the tables with user-defined partitions. Thanks!